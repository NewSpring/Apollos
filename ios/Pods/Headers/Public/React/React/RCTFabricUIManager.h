../../../../../../node_modules/react-native/React/Fabric/RCTFabricUIManager.h