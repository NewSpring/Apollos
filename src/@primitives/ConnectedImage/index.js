import React, { PureComponent } from 'react';
import { Platform } from 'react-native';
import { every } from 'lodash';
import PropTypes from 'prop-types';
import { Image } from 'react-native';

// This mirrors the File resource we get from Heighliner:
const ImageSourceType = PropTypes.oneOfType([
  PropTypes.shape({
    url: PropTypes.string,
    label: PropTypes.string,
  }),
  PropTypes.string,
]);

const sizeCache = {};

const getCachedSources = (_sources = []) => {
  let sources = _sources;
  if (!Array.isArray(sources)) sources = [sources];
  sources = sources.map((source) => {
    if (typeof source === 'string') return { url: source };
    return source;
  });

  return sources.map(source => ({
    uri: (source.url || '').replace(/^http:\/\/|^\/\//i, 'https://'),
    ...source,
    ...(sizeCache[source.size] || {}),
  }));
};

class ConnectedImage extends PureComponent {
  static propTypes = {
    source: PropTypes.oneOfType([
      PropTypes.arrayOf(ImageSourceType), ImageSourceType,
    ]),
    ImageComponent: PropTypes.oneOfType([PropTypes.func, PropTypes.object]),
  }

  static defaultProps = {
    ImageComponent: Image,
  }

  state = {
    source: getCachedSources(this.props.source),
  };

  componentWillMount() { this.updateCache(this.props.source); }
  componentWillReceiveProps(newProps) { this.updateCache(newProps.source); }

  async updateCache(sources) {
    await Promise.all(getCachedSources(sources).map((source) => {
      if (sizeCache[source.size]) return Promise.resolve(source);
      return (new Promise((resolve, reject) => {
        Image.getSize(source.uri, (width, height) => resolve({
          width, height,
        }), reject);
      })).then((sizeForCache) => {
        sizeCache[source.size] = sizeForCache;
      });
    }));

    this.setState({ source: getCachedSources(sources) });
  }

  render() {
    let { source } = this.state;
    // Android can't currently render an image source without a width/height specified, and then
    // re-render that source with width/height. So render null until width and height is set:
    if (Platform.OS === 'android') {
      if (!Array.isArray(source)) source = [source];
      if (!every(this.state.source, image => image.width && image.height)) return null;
    }

    // react-native-web currently doesn't support array-based Image sources
    if (Platform.OS === 'web' && Array.isArray(source)) {
      [source] = source;
    }

    const { ImageComponent = Image, ...otherProps } = this.props;
    return <ImageComponent {...otherProps} source={source} />;
  }
}

export default ConnectedImage;
