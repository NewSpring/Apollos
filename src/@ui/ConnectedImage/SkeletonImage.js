import React from 'react';
import { Platform, View } from 'react-native';
import { compose } from 'recompose';

import Placeholder from 'rn-placeholder';
import styled from '@ui/styled';

const enhance = compose(
  Placeholder.connect,
);

const StyledView = styled(({ theme }) => ({
  width: '100%',
  aspectRatio: 1,
  backgroundColor: theme.colors.background.inactive,
  ...Platform.select({
    web: {
      // web doesn't support aspectRatio, this hacks it:
      height: 0,
      paddingTop: '100%',
    },
  }),
}))(View);

const SkeletonImage = enhance(() => (
  <StyledView />
));

export default SkeletonImage;
