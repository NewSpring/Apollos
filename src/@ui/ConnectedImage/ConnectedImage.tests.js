import React from 'react';
import renderer from 'react-test-renderer';

import ConnectedImage, {
  getCachedSources,
  updateCache,
} from './';

describe('the ConnectedImage component', () => {
  it('should render immediately with a network image with a known width and height', () => {
    const tree = renderer.create(
      <ConnectedImage
        source={{
          uri: 'https://placeholdit.co/i/150x150?bg=eeeeee&fc=577084',
          width: 150,
          height: 150,
        }}
      />,
    );
    expect(tree).toMatchSnapshot();
  });
  it('should maintain aspect ratio', () => {
    const tree = renderer.create(
      <ConnectedImage
        source={{
          uri: 'https://placeholdit.co/i/150x150?bg=eeeeee&fc=577084',
          width: 150,
          height: 150,
        }}
        maintainAspectRatio
      />,
    );
    expect(tree).toMatchSnapshot();
  });
  describe('updateCache', () => {
    it('updates cache with image uri and sizes', async () => {
      const source = {
        url: '//via.placeholder.com/320x240',
      };
      await updateCache(source);

      expect(getCachedSources(source)).toContainEqual(expect.objectContaining({
        uri: 'https://via.placeholder.com/320x240',
        url: '//via.placeholder.com/320x240',
        width: 320,
        height: 240,
      }));
    });
  });
});
