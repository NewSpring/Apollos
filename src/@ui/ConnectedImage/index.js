import React, { PureComponent } from 'react';
import { Platform, Image } from 'react-native';
import PropTypes from 'prop-types';
import { every } from 'lodash';

import SkeletonImage from './SkeletonImage';

// This mirrors the File resource we get from Heighliner:
const ImageSourceType = PropTypes.shape({
  uri: PropTypes.string,
  label: PropTypes.string,
  width: PropTypes.number,
  height: PropTypes.number,
});

export const sizeCache = {};

export const getCacheKey = (source) => {
  if (source.size && source.fileLabel) return `${source.size}-${source.fileLabel}`;
  if (source.url) return source.url;
  if (source.uri) return source.uri;
  return undefined;
};

export const getCachedSources = (_sources = []) => {
  let sources = _sources;
  if (!Array.isArray(sources)) sources = [sources];
  sources = sources.map((source) => {
    if (typeof source === 'string') return { url: source };
    return source;
  });

  return sources.map(source => ({
    uri: (source.url || '').replace(/^http:\/\/|^\/\//i, 'https://'),
    ...source,
    ...(sizeCache[getCacheKey(source)] || {}),
  }));
};

export const updateCache = sources => Promise.all(getCachedSources(sources).map((source) => {
  const key = getCacheKey(source);
  if (sizeCache[key] || !key) return Promise.resolve(source);
  return (new Promise((resolve, reject) => {
    Image.getSize(source.uri, (width, height) => resolve({
      width, height,
    }), reject);
  })).then((sizeForCache) => {
    if (key) sizeCache[key] = sizeForCache;
  });
}));

class ConnectedImage extends PureComponent {
  static propTypes = {
    source: PropTypes.oneOfType([
      PropTypes.arrayOf(ImageSourceType),
      ImageSourceType,
    ]),
    ImageComponent: PropTypes.oneOfType([PropTypes.func, PropTypes.object]),
    maintainAspectRatio: PropTypes.bool,
    style: PropTypes.any, // eslint-disable-line
  }

  static defaultProps = {
    ImageComponent: Image,
    maintainAspectRatio: false,
  }

  state = {
    source: getCachedSources(this.props.source),
  };

  componentWillMount() { this.updateCache(this.props.source); }
  componentWillReceiveProps(newProps) { this.updateCache(newProps.source); }

  get aspectRatio() {
    const style = {};
    if (this.props.maintainAspectRatio) {
      const firstSource = this.state.source[0];
      if (firstSource && firstSource.width && firstSource.height) {
        style.aspectRatio = firstSource.width / firstSource.height;
        if (Platform.OS === 'web') {
          style.height = 0;
          style.paddingTop = `${style.aspectRatio * 100}%`;
        }
      }
    }
    return style;
  }

  get isLoading() {
    return !every(this.state.source, image => image.width && image.height);
  }

  async updateCache(sources) {
    await updateCache(sources);
    this.setState({ source: getCachedSources(sources) });
  }

  render() {
    let { source } = this.state;
    if (!Array.isArray(source)) source = [source];

    // react-native-web currently doesn't support array-based Image sources
    if (Platform.OS === 'web' && Array.isArray(source)) {
      [source] = source;
    }

    const { ImageComponent = Image, style, ...otherProps } = this.props;

    return (
      <SkeletonImage onReady={!this.isLoading} animate={'fade'}>
        <ImageComponent {...otherProps} source={source} style={[this.aspectRatio, style]} />
      </SkeletonImage>
    );
  }
}

export default ConnectedImage;
