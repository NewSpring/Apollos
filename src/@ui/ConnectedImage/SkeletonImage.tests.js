import React from 'react';
import renderer from 'react-test-renderer';
import { ThemeProvider } from '@ui/theme';

import SkeletonImage from './SkeletonImage';

describe('the SkeletonImage component', () => {
  it('should render a loading state', () => {
    const tree = renderer.create(
      <ThemeProvider>
        <SkeletonImage onReady={false} />
      </ThemeProvider>,
    );
    expect(tree).toMatchSnapshot();
  });
});
