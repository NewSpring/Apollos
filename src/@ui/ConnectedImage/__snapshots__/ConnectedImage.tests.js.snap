// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`the ConnectedImage component should maintain aspect ratio 1`] = `
<Image
  collapsable={undefined}
  onLoad={[Function]}
  source={
    Array [
      Object {
        "height": 150,
        "uri": "https://placeholdit.co/i/150x150?bg=eeeeee&fc=577084",
        "width": 150,
      },
    ]
  }
  style={
    Object {
      "aspectRatio": 1,
      "backgroundColor": "#dddddd",
      "opacity": 0,
    }
  }
/>
`;

exports[`the ConnectedImage component should render immediately with a network image with a known width and height 1`] = `
<Image
  collapsable={undefined}
  onLoad={[Function]}
  source={
    Array [
      Object {
        "height": 150,
        "uri": "https://placeholdit.co/i/150x150?bg=eeeeee&fc=577084",
        "width": 150,
      },
    ]
  }
  style={
    Object {
      "backgroundColor": "#dddddd",
      "opacity": 0,
    }
  }
/>
`;
